### PR TITLE
Issue 1401: Add a generic end-point for deleting repository paths in the StoragesConfigurationController

### LINKED ISSUE #1401: Add a generic end-point for deleting repository paths in the StoragesConfigurationController
# Task Description

We need to create an end-points for moving repository paths. The code should be located in the `StoragesConfigurationController` class and a separate form class should be created so that it is able to handle the `JSON` below. Be sure to check our [How to write Spring Controllers guide](https://github.com/strongbox/strongbox/wiki/How-To-Implement-Spring-Controllers) to see how we return errors. There is some partial implementation of this in the `MavenArtifactController` which could be useful to speed-up the development of this feature. The existing implementation should be refactored as much as possible and made generic enough that it works for all layout providers.

### Action:

Deletes recursively the content of `src.path

### Form validation:

1. `src*` are all required.

```json
DELETE /api/configuration/strongbox/storages/delete

{
   'srcStorageId': 'storage0',
   'srcRepositoryId': 'some-repository-id',
   'srcPath': 'path/relative/to/the/a/directory/or/file',
}
```

# Tasks

The following tasks will need to be carried out:
* [ ] Implement end-point for deleting paths
* [ ] Add test cases checking the above for the following layout providers:
  * [ ] Maven
  * [ ] NuGet
  * [ ] Raw
  * [ ] PyPi 

# Task Relationships

This task relates to: #1399, #1400

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @fuss86
  * @sbespalov
  * @steve-todorov

### PR DESCRIPTION
**Pull request by mans4321**
# Pull Request Description

This pull request closes #1401 .

# Acceptance Test

* [x] Building the code with `mvn clean install -Dintegration.tests` still works.
* [x] Running `mvn spring-boot:run` in the `strongbox-web-core` still starts up the application correctly.
* [x] Building the code and running the `strongbox-distribution` from a `zip` or `tar.gz` still works.
* [x] The tests in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/) still run properly.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [x] No

* Does this require an update of the documentation?
  * [ ] Yes, please see strongbox/strongbox-docs#{PR_NUMBER}
  * [x] No

### CODE REVIEW

**Comment by carlspring** [comment]
Hi!

Thank you for your contribution!

Would you mind signing the ICLA, as described in the [Contributing](https://github.com/strongbox/strongbox/blob/master/CONTRIBUTING.md) page?

Also, please, feel free to join our [chat channel](https://chat.carlspring.org/), if you'd like to learn more about the project and/or like to find out what else you could help with.

Kind regards,

Martin

**Comment by mans4321** [comment]
Sure I will do that on Friday. Thanks for the feedback.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/forms/PathForm.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
+package org.carlspring.strongbox.forms;
+
+import javax.validation.constraints.NotEmpty;
+import javax.validation.constraints.Pattern;
+
+public class PathForm {
+
+    @NotEmpty(message = "A storage id must be specified.")
+    @Pattern(regexp = "[a-zA-Z0-9\\-\\_\\.]+")
```
Comment (ptirador) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/forms/PathForm.java]: Redundant character escapes for `_` and `.` inside square brackets.
```suggestion
    @Pattern(regexp = "[a-zA-Z0-9\\-_.]+")
```

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/forms/PathForm.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    private String storageId;
+
+    @NotEmpty(message = "An repository id must be specified.")
+    @Pattern(regexp = "[a-zA-Z0-9\\-\\_\\.]+")
```
Comment (ptirador) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/forms/PathForm.java]: Redundant character escapes for `_` and `.` inside square brackets.
```suggestion
    @Pattern(regexp = "[a-zA-Z0-9\\-_.]+")
```

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationController.java [hunk 3]
```diff
@@ -332,4 +345,35 @@ public ResponseEntity removeRepository(@RepositoryMapping Repository repository,
...
+            MediaType.APPLICATION_JSON_VALUE })
+    public ResponseEntity removePath(@ApiParam(value = "The path object", required = true)
+                                     @RequestBody @Validated({ Default.class,
+                                                               ProxyConfigurationFormChecks.class})
+                                             PathForm pathForm,
+                                     BindingResult bindingResult,
+                                     @RequestHeader(HttpHeaders.ACCEPT) String accept) {
+
+        if(bindingResult.hasErrors())
+            return getFailedResponseEntity(HttpStatus.NOT_FOUND, FAILED_PATH_REMOVAL_FORM_ERROR, accept);
```
Comment (ptirador) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationController.java]: Please, could you use braces around the `if` block?

**Comment by carlspring** [comment]
Hi @mans4321 ,

Thanks ofr signing the ICLA! :)

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationControllerTestIT.java [hunk 4]
```diff
@@ -667,4 +696,96 @@ public void whenStorageIsCreatedWithoutBasedirProvidedDefaultIsSet(String extens
...
+        PathForm pathForm = buildPathForm(repository.getStorage().getId(), repository.getId(), repositoryPathStr, false);
+
+        givenCustom().contentType(MediaType.APPLICATION_JSON_VALUE)
+                .accept(MediaType.APPLICATION_JSON_VALUE)
+                .body(pathForm)
+                .when()
+                .delete(url)
+                .prettyPeek()
+                .then()
+                .body(containsString(SUCCESSFUL_PATH_REMOVAL));
```
Comment (carlspring) [inline strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationControllerTestIT.java]: Cosmetic: Could you please align chained methods in accordance with our [coding convention](https://strongbox.github.io/developer-guide/coding-convention.html) by the first dot like this:

```
        givenCustom().contentType(MediaType.APPLICATION_JSON_VALUE)
                     .accept(MediaType.APPLICATION_JSON_VALUE)
                     .body(pathForm)
                     .when()
                     .delete(url)
                     .prettyPeek()
                     .then()
                     .body(containsString(SUCCESSFUL_PATH_REMOVAL));
```

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/forms/PathForm.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        return storageId;
+    }
+
+    public String getRepositoryId()
+    {
+        return repositoryId;
+    }
+
+    public String getPath() {
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/forms/PathForm.java]: Cosmetic: Braces on new lines, please. Thanks! :)

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationControllerTestIT.java [hunk 6]
```diff
@@ -667,4 +696,96 @@ public void whenStorageIsCreatedWithoutBasedirProvidedDefaultIsSet(String extens
...
+    @ExtendWith({ RepositoryManagementTestExecutionListener.class,
+                  ArtifactManagementTestExecutionListener.class })
+    @Test
+    public void testDeleteNpmPathInRepository(@NpmRepository(repositoryId = NPM_REPOSITORY_RELEASES)
+                                                             Repository repository,
+                                              @NpmTestArtifact(repositoryId = NPM_REPOSITORY_RELEASES,
+                                                               versions = "1.0.0",
+                                                               id = "npm-test-release",
+                                                               scope = "@carlspring")
+                                              Path path)
```
Comment (carlspring) [inline strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationControllerTestIT.java]: Cosmetic: Alignment:
```
    public void testDeleteNpmPathInRepository(@NpmRepository(repositoryId = NPM_REPOSITORY_RELEASES)
                                              Repository repository,
                                              @NpmTestArtifact(repositoryId = NPM_REPOSITORY_RELEASES,
                                                               versions = "1.0.0",
                                                               id = "npm-test-release",
                                                               scope = "@carlspring")
                                              Path path)
```

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationController.java [hunk 7]
```diff
@@ -332,4 +345,42 @@ public ResponseEntity removeRepository(@RepositoryMapping Repository repository,
...
+        }
+
+        try
+        {
+            final RepositoryPath repositoryPath = repositoryPathResolver.resolve(pathForm.getStorageId(),
+                    pathForm.getRepositoryId(),
+                    pathForm.getPath());
+            
+            artifactManagementService.delete(repositoryPath, pathForm.isForce());
+            logger.info("Deleting Path {}:{}/{}...", pathForm.getStorageId(), pathForm.getRepositoryId(), pathForm.getPath());
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationController.java]: Cosmetic: Could you keep `logger` blocks separated by a line before and after? This makes it a bit more readable. Thanks! :)

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationController.java [hunk 8]
```diff
@@ -332,4 +345,42 @@ public ResponseEntity removeRepository(@RepositoryMapping Repository repository,
...
+        if(bindingResult.hasErrors())
+        {
+            return getFailedResponseEntity(HttpStatus.NOT_FOUND, FAILED_PATH_REMOVAL_FORM_ERROR, accept);
+        }
+
+        try
+        {
+            final RepositoryPath repositoryPath = repositoryPathResolver.resolve(pathForm.getStorageId(),
+                    pathForm.getRepositoryId(),
+                    pathForm.getPath());
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationController.java]: Cosmetic: Please, re-align the method parameters like this:
```
            final RepositoryPath repositoryPath = repositoryPathResolver.resolve(pathForm.getStorageId(),
                                                                                 pathForm.getRepositoryId(),
                                                                                 pathForm.getPath());
```

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationControllerTestIT.java [hunk 9]
```diff
@@ -667,4 +696,96 @@ public void whenStorageIsCreatedWithoutBasedirProvidedDefaultIsSet(String extens
...
+
+    private void deleteRepositoryPath(Repository repository, Path path)
+            throws IOException
+    {
+        String url = getContextBaseUrl() + "/delete";
+
+        RepositoryPath repositoryPath = (RepositoryPath) path.toRealPath();
+        String repositoryPathStr = RepositoryFiles.relativizePath(repositoryPath);
+
+        PathForm pathForm = buildPathForm(repository.getStorage().getId(), repository.getId(), repositoryPathStr, false);
```
Comment (carlspring) [inline strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationControllerTestIT.java]: Cosmetic: Could you please, chop these parameters as one per line and indent them properly? Thanks! :)

**Comment by carlspring** [comment]
@ptirador : Are you happy with this pull request?

**Comment by carlspring** [comment]
@mans4321 ,

Let me know when you've had a chance to address the change requests and I'll re-review and merge it, (unless @ptirador has any other remarks (...?)).

Thanks! :)

**Comment by mans4321** [comment]
@ptirador  to answer your question I wasn't before. however, in the new commit I add a few lines to improve the code quality .

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationController.java [hunk 10]
```diff
@@ -349,6 +349,7 @@ public ResponseEntity removeRepository(@RepositoryMapping Repository repository,
     @ApiResponses(value = { @ApiResponse(code = 200, message = "The path was deleted successfully."),
                             @ApiResponse(code = 404, message = "The path cannot be deleted because storageId,repositoryId,or path was not found!"),
                             @ApiResponse(code = 500, message = "Failed to remove the path") })
+    @PreAuthorize("hasAuthority('CONFIGURATION_DELETE_REPOSITORY_PATH')")
```
Comment (mans4321) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationController.java]: before I was missing PreAuthorize.
@ptirador

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationControllerTestIT.java [hunk 11]
```diff
@@ -762,19 +762,30 @@ private void deleteRepositoryPath(Repository repository, Path path)
...
+                     .body(pathForm)
+                     .when()
+                     .delete(url)
+                     .prettyPeek()
+                     .then()
+                     .body(containsString(SUCCESSFUL_PATH_REMOVAL));
+
+        assertThat(Files.notExists(path))
+                .as("Failed to delete repository path '" + path + "'!")
+                .isTrue();
```
Comment (mans4321) [inline strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationControllerTestIT.java]: also I added this
@ptirador

**Comment by carlspring** [comment]
@mans4321 ,

Could you please rebase and fix the failing tests? Thanks! :)

**Comment by mans4321** [comment]
I am really Sorry this PR got really dirty so I decided to open to make everyone life easy 
https://github.com/strongbox/strongbox/pull/1592

**Comment by steve-todorov** [comment]
Thanks for the new PR @mans4321 - will be closing this one in that case.

**Pull request closed without merging**